src/openstack_cloud/openstack_manager.py

```diff
@@ -0,0 +1,378 @@
+    path: GithubPath,
+    proxies: Optional[ProxyConfig] = None,
+) -> str:
+    """Build and upload an image to OpenStack.
+
+    Args:
+        cloud_config: The cloud configuration to connect OpenStack with.
+        github_client: The Github client to interact with Github API.
+        path: Github organisation or repository path.
+        proxies: HTTP proxy settings.
+
+    Raises:
+        ImageBuildError: If there were errors building/creating the image.
+
+    Returns:
+        The created OpenStack image id.
+    """
+    try:
+        runner_application = github_client.get_runner_application(path=path, arch=arch)
+    except RunnerBinaryError as exc:
+        raise OpenstackImageBuildError("Failed to fetch runner application.") from exc
+
+    try:
+        execute_command(_build_image_command(runner_application, proxies), check_exit=True)
+    except SubprocessError as exc:
+        raise OpenstackImageBuildError("Failed to build image.") from exc
+
+    try:
+        runner_arch = runner_application["architecture"]
+        image_arch = _get_supported_runner_arch(arch=runner_arch)
+    except UnsupportedArchitectureError as exc:
+        raise OpenstackImageBuildError(f"Unsupported architecture {runner_arch}") from exc
+
+    try:
+        conn = _create_connection(cloud_config)
+        existing_image: openstack.image.v2.image.Image
+        for existing_image in conn.search_images(name_or_id=IMAGE_NAME):
+            # images with same name (different ID) can be created and will error during server
+            # instantiation.
+            if not conn.delete_image(name_or_id=existing_image.id, wait=True):
+                raise OpenstackImageBuildError("Failed to delete duplicate image on Openstack.")
+        image: openstack.image.v2.image.Image = conn.create_image(
+            name=IMAGE_NAME, filename=IMAGE_PATH_TMPL.format(architecture=image_arch), wait=True
+        )
+        return image.id
+    except OpenStackCloudException as exc:
+        raise OpenstackImageBuildError("Failed to upload image.") from exc
+
+
+def create_instance_config(
+    unit_name: str,
+    openstack_image: openstack.image.v2.image.Image,
+    path: GithubPath,
+    github_client: GithubClient,
+) -> InstanceConfig:
+    """Create an instance config from charm data.
+
+    Args:
+        unit_name: The charm unit name.
+        image: Ubuntu image flavor.
+        path: Github organisation or repository path.
+        github_client: The Github client to interact with Github API.
+    """
+    app_name, unit_num = unit_name.rsplit("/", 1)
+    suffix = secrets.token_hex(12)
+    registration_token = github_client.get_runner_registration_token(path=path)
+    return InstanceConfig(
+        name=f"{app_name}-{unit_num}-{suffix}",
+        labels=(app_name, "jammy"),
+        registration_token=registration_token,
+        github_path=path,
+        openstack_image=openstack_image,
+    )
+
+
+class InstanceLaunchError(Exception):
```

Comment:
I think "public" exceptions are generally defined in the `errors.py` file.
By "public", I am referring to exception that would be import and caught in another module, e.g., `charm.py`.

It seems this exception would be caught by another module?

If this is moved to `errors.py`, it might be good to rename with Openstack in the name.

Reply:
Moved, thank you!